### DISCUSSION TITLE
To give a configration option for allowing case insensitive URLS

### DISCUSSION THREAD

**Discussion by mufaddal7** [vercel/next.js#21733] [Ideas]
**Describe the feature you'd like to request**

URLs with uppercase and lowercase should result in the same page and not throw 404.
*For example*
- /Product-Name
- /product-name
- /PRODUCT-NAME 
all should result the same page.

We can give a nextjs configuration in **next.config.js** file

**Describe the solution you'd like**

There can be different methods to implement this, I could imagine a solution like this 

next.config.js can have an option for case sensitivity
```
module.exports = {
  options: {
    case:'insentive'
  },
}
```
**Describe alternatives you've considered**

We can create something like this.
```
 module.exports = {
    async URLoptions() {
      return [
        {
          case: 'insensitive'
        },
      ]
    },
  }
```
There are many issues raising regarding case-sensitive URLs, this could be beneficial for many developers. Especially when there is a migration from another stack where the URL case didn't affect the webpage.

**Comment by cobrexus** [comment]
This is not optimal, but if really necessary, you could use Next.js [catch all routes](https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes).

### If your application has complex dynamic routes (e.g. `/a/b/c`):

Place the below code in `pages/[...route].js`:

```jsx
// pages/[...route].js

import { useEffect } from 'react'
import { useRouter } from 'next/router'

export default function ResolveRoute() {
    const router = useRouter()

    useEffect(() => {
        const lowerCaseRoutes = router.query.route.map((param) => {
            return `/${param.toLowerCase()}`
        })
        const newRoute = lowerCaseRoutes.toString().replaceAll(',', '')
        router.push(newRoute)
    })

    return <p>Redirecting...</p>
}
```

### If your application has only simple routes (e.g. `/a`, `/b`):

Place the below code in `pages/[route].js`:

```jsx
// pages/[route].js

import { useEffect } from 'react'
import { useRouter } from 'next/router'

export default function ResolveRoute() {
    const router = useRouter()

    useEffect(() => {
        router.push(router.pathname.toLowerCase())
    })

    return <p>Redirecting...</p>
}
```

**Reply by bcanedo4** [reply]
I would also check if the router.pathname is not equal to the router.pathname.toLowerCase() before calling router.push(), to avoid unnecessary calls:
`useEffect(() => {
  if (router.pathname !== router.pathname.toLowerCase()) {
    router.push(router.pathname.toLowerCase())
  }
})`
